### PR TITLE
Implement IDisposable pattern and release resources on Dispose.

### LINKED ISSUE #439: Implement IDisposable pattern
Throughout the code base the implementation of `IDisposable` does not follow [the recommended guidelines](https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose "Implementing a Dispose method").

### PR DESCRIPTION
**Pull request by paulomorgado**
## Proposed Changes

Throughout the code base the implementation of `IDisposable` does not follow [the recommended guidelines](https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose "Implementing a Dispose method").

Fixes #439

## Types of Changes

- [ ] Bugfix (non-breaking change which fixes issue #NNNN)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (correction or otherwise)
- [ ] Cosmetics (whitespace, appearance)

## Checklist

- [x] I have read the `CONTRIBUTING.md` document
- [x] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [ ] All tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in related repositories

## Further Comments

This is more of a bootstrap for the correct implementation as I haven't check the implications on derived classes.

### CODE REVIEW

File: projects/client/RabbitMQ.Client/src/client/impl/AutorecoveringConnection.cs [hunk 1]
```diff
@@ -183,15 +189,15 @@ private bool ManuallyClosed
             {
                 lock (m_eventLock)
                 {
-                    m_recordedBlockedEventHandlers.Add(value);
+                    m_recordedBlockedEventHandlers += value;
```
Comment (michaelklishin) [inline projects/client/RabbitMQ.Client/src/client/impl/AutorecoveringConnection.cs]: How is this relevant to the goal of this PR?

**Comment by michaelklishin** [comment]
@bording can you think of anything potentially breaking here?

File: projects/client/RabbitMQ.Client/src/client/impl/AutorecoveringConnection.cs (hunk 1 above)
Comment (paulomorgado) [reply projects/client/RabbitMQ.Client/src/client/impl/AutorecoveringConnection.cs]: Resource usage. Lists of event handlers are being created regardless of being used or not. And for the purpose of what a multicast delegate can do.

**Comment by michaelklishin** [comment]
Thank you!

**Comment by lukebakken** [comment]
Thanks @paulomorgado - does this PR fix [this issue](https://groups.google.com/d/topic/rabbitmq-users/aAkVIBmoaC4/discussion) that you reported to `rabbitmq-users`?

File: projects/client/RabbitMQ.Client/src/client/impl/AutorecoveringConnection.cs [hunk 2]
```diff
@@ -76,8 +76,7 @@ public class AutorecoveringConnection : IConnection, IRecoverable
         protected ConcurrentDictionary<RecordedBinding, byte> m_recordedBindings =
             new ConcurrentDictionary<RecordedBinding, byte>();
 
-        protected List<EventHandler<ConnectionBlockedEventArgs>> m_recordedBlockedEventHandlers =
-            new List<EventHandler<ConnectionBlockedEventArgs>>();
+        protected EventHandler<ConnectionBlockedEventArgs> m_recordedBlockedEventHandlers;
```
Comment (bording) [inline projects/client/RabbitMQ.Client/src/client/impl/AutorecoveringConnection.cs]: Changing the type of protected member would be a breaking change to any derived types. `AutorecoveringConnection` is a public type, so if you're strictly avoiding breaks, this shouldn't be changed.

File: projects/client/RabbitMQ.Client/src/client/impl/AutorecoveringConnection.cs [hunk 3]
```diff
@@ -106,6 +103,15 @@ public AutorecoveringConnection(ConnectionFactory factory, string clientProvided
             this.ClientProvidedName = clientProvidedName;
         }
 
+        /// <summary>
+        /// Finalizes an instance of the <see cref="AutorecoveringConnection"/> class.
+        /// </summary>
+        ~AutorecoveringConnection()
```
Comment (bording) [inline projects/client/RabbitMQ.Client/src/client/impl/AutorecoveringConnection.cs]: There is no reason to implement finalizers on any of these types. This is only something you need to do if you have unmanaged resources that need have to be freed even when `Dispose` hasn't been called.

**Comment by bording** [comment]
@michaelklishin All of the event handler changes are potentially breaking, and there is no need for the finalizers.

I'm not sure the changes made here actually make any real impact.

**Comment by michaelklishin** [comment]
Thank you, @bording. Let's wait for evidence that this makes a difference for the workload linked to above (the rabbitmq-users thread).

File: projects/client/RabbitMQ.Client/src/client/impl/AutorecoveringConnection.cs (hunk 2 above)
Comment (michaelklishin) [reply projects/client/RabbitMQ.Client/src/client/impl/AutorecoveringConnection.cs]: I am not aware of any projects that would derive from this class. Does NServiceBus do it?

Comment (bording) [reply projects/client/RabbitMQ.Client/src/client/impl/AutorecoveringConnection.cs]: No, we don't. I'm guessing the odds are low that anyone does, but those types are `protected`, which means that someone could observe the change, and you'd break them, though.

File: projects/client/RabbitMQ.Client/src/client/impl/AutorecoveringConnection.cs (hunk 3 above)
Comment (lukebakken) [reply projects/client/RabbitMQ.Client/src/client/impl/AutorecoveringConnection.cs]: @bording is correct, and [the provided link backs it up](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/dispose-pattern#finalizable-types).

Comment (paulomorgado) [reply projects/client/RabbitMQ.Client/src/client/impl/AutorecoveringConnection.cs]: How come? Doesn't `CreateConnection` create a `Socket` instance?

Of course this is broken because `IProtocol` doesn't implement `IDisposable`. But we have to start somewhere.

Comment (bording) [reply projects/client/RabbitMQ.Client/src/client/impl/AutorecoveringConnection.cs]: `Socket` is still a managed resource, so just call `Dispose`, and it will handle any native OS cleanup.

You only need a finalizer when you have directly allocated native resources that won't be freed if you don't do it yourself.

Comment (paulomorgado) [reply projects/client/RabbitMQ.Client/src/client/impl/AutorecoveringConnection.cs]: @bording 

> You only need a finalizer when you have directly allocated native resources that won't be freed if you don't do it yourself.

Fair enough!

**Comment by kjnilsson** [comment]
Thanks. Is there any value in leaving the `GC.SuppressFinalize` calls in there?

File: projects/client/RabbitMQ.Client/src/client/impl/AutorecoveringConnection.cs (hunk 2 above)
Comment (kjnilsson) [reply projects/client/RabbitMQ.Client/src/client/impl/AutorecoveringConnection.cs]: these can all be made private. Now that work on 6.0 has started I would happily accept PRs that hides fields and data that ought to be internal.

**Comment by michaelklishin** [comment]
@paulomorgado sorry, any comments on `GC.SuppressFinalize` calls above? I'm not sure of the state of this PR. We'd prefer to close PRs that are not making clear progress.

**Comment by lukebakken** [comment]
> Is there any value in leaving the GC.SuppressFinalize calls in there?

Yes, that is part of the standard implementation of `IDisposable` as noted in the reference docs.

**Comment by kjnilsson** [comment]
The thing is that `GC.SuppressFinalize` doesn't do anything unless there is a finalizer other than the default finalizer implementation on `Object` so in this case I don't see it adding anything, unless I am missing something.

**Comment by lukebakken** [comment]
There is an interesting comment [here](https://docs.microsoft.com/en-us/dotnet/api/system.idisposable?view=netframework-4.7.2):

```
        // Implement IDisposable.
        // Do not make this method virtual.
        // A derived class should not be able to override this method.
        public void Dispose()
        {
            Dispose(true);
            // This object will be cleaned up by the Dispose method.
            // Therefore, you should call GC.SupressFinalize to
            // take this object off the finalization queue
            // and prevent finalization code for this object
            // from executing a second time.
            GC.SuppressFinalize(this);
        }
```

It looks like there is a valid reason for it and, if it's not there, I think people will be surprised should they review the code.

**Comment by bording** [comment]
@lukebakken In that example, the class has a finalizer defined, so it will be put on the finalization queue. If the class doesn't have a finalizer at all, it won't be put on the queue in the first place, so as @kjnilsson said,  calling `GC.SuppressFinalize` doesn't do anything.

**Comment by paulomorgado** [comment]
The classes are open to derivation (that's another issue). Any derived class can have a finalizer. The first example of the [Basic Dispose Pattern](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/dispose-pattern#basic-dispose-pattern) is an example of that.

**Comment by kjnilsson** [comment]
Merged without GC.SuppressFinalize. Thanks!

**Comment by paulomorgado** [comment]
I just want to make it clear that this is an issue if someone overrides any of these classes.
